chan1m32isis gtest

AustinSanders opened the pull request:
Converts chan1m32isis to .cpp/.h and converts makefile tests to gtest

## Description
Converts chan1m32isis to .cpp/.h and converts makefile tests to gtest

## Related Issue
Test conversion

## Motivation and Context
Eliminates 1176Mb of test data

## How Has This Been Tested?
6 Gtests

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- - [ ] My code follows the code style of this project. -->
- [x] I have read and agree to abide by the [Code of Conduct](https://usgs-astrogeology.github.io/code/)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [x] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

AustinSanders committed:
Initial constants unit gtest

AustinSanders committed:
Merge branch 'dev' of https://github.com/USGS-Astrogeology/ISIS2 into dev

AustinSanders committed:
Convert chan1m32isis for gtest

AustinSanders commented on isis/src/chandrayaan1/apps/chan1m32isis/chan1m32isis.cpp:
```diff
@@ -0,0 +1,480 @@
+  PvlGroup g_results("Results");
+
+  void chan1m32isis(UserInterface &ui) {
+    //g_results.clear();
```
This line results in a segfault during testing.  This needs to be fixed, but I wanted to get the PR up ASAP.  Any ideas?

kaitlyndlee replied:
UserInterface inherits from IsisAml and the `WasEntered` method called in `importImage` is inherited from IsisAml. This could be relating to the problem and the fact that in the tests, we are not using the Application class anymore.

AustinSanders commented on isis/tests/FunctionalTestsChan1m32isis.cpp:
```diff
@@ -0,0 +1,550 @@
+}
+
+
+TEST(Chan1m32Isis, Chan1m32IsisTestFowardDescending) {
```
I'm not sure if we need to test all of the same keywords + DNs for each of the forward/reverse ascending/descending, but I wrote the tests anyway.  We can trim this down as needed.

kaitlyndlee commented on isis/tests/data/chan1m32isis/badFile/bad.file:
```diff
@@ -0,0 +1 @@
+>:(
```
Is this needed or could we use an ofstream to create it programmatically? An example of this is in the campt tests.

kaitlyndlee commented on isis/tests/FunctionalTestsChan1m32isis.cpp:
```diff
@@ -0,0 +1,550 @@
+                           "to=" + cubeFileName };
+
+  UserInterface options(APP_XML, args);
+  ASSERT_THROW(chan1m32isis(options), IException);
```
@Kelvinrr It seems like there are multiple ways we are checking errors. I think we should come up with a standard.

Kelvinrr replied:
Good catch.

I think sticking with `try{app(); FAIL();}Catch(){e.what substring check}` is the way to go, it allows you to assert both the error type and the message. Especially for cases like this where the exception type is not enough to reasonably assume that the Exception was thrown for the right reason. Here, IException can be triggered by a number of things that may not be what you want. 

@AustinSanders I would recommend switching to the try/except approach instead. see: https://github.com/USGS-Astrogeology/ISIS3/blob/24b1fc7e1c2c1182f2b454c4ecb0d4394a424d14/isis/tests/FunctionalTestsJigsaw.cpp#L422

AustinSanders committed:
Points badfile test to existing product

AustinSanders committed:
Refactored for logging

acpaquette commented on isis/tests/FunctionalTestsChan1m32isis.cpp:
```diff
@@ -0,0 +1,558 @@
+  QTemporaryDir prefix;
+  QString cubeFileName = prefix.path() + "/chan1m32isisTEMP.cub";
+  QVector<QString> args = {"from=data/chan1m32isis/forwardAscending/M3T20090630T083407_V03_L1B_cropped.LBL",
+                           "loc=data/chan1m32isis/forwardAscending/M3T20090630T083407_V03_LOC_cropped.IMG",
+                           "obs=data/chan1m32isis/forwardAscending/M3T20090630T083407_V03_OBS_cropped.IMG",
```
Could we copy these into the temporary testing directory that gets created and use the copied files here. If we don't there are a couple cubes that get created in the test data directory. Maybe using something like https://en.cppreference.com/w/cpp/filesystem/copy

acpaquette commented on isis/tests/FunctionalTestsChan1m32isis.cpp:
```diff
@@ -0,0 +1,558 @@
+
+  // Dimensions Group
+  PvlGroup &dimensions = isisLabel->findGroup("Dimensions", Pvl::Traverse);
+  ASSERT_EQ(int(dimensions["Samples"]), 608);
+  ASSERT_EQ(int(dimensions["Lines"]), 5);
+  ASSERT_EQ(int(dimensions["Bands"]), 3);
+
+  // Pixels Group
+  PvlGroup &pixels = isisLabel->findGroup("Pixels", Pvl::Traverse);
+  ASSERT_EQ(pixels["Type"][0].toStdString(), "Real");
+  ASSERT_EQ(pixels["ByteOrder"][0].toStdString(), "Lsb");
+  ASSERT_EQ(double(pixels["Base"]), 0.0);
+  ASSERT_EQ(double(pixels["Multiplier"]), 1.0);
```
Another comment made on another 2isis PR, these values can be accessed directly from functions that exist on the cube. 

`sample/line/bandcount` for the number of samples, lines, and bands
`pixelType, bandOrder, base, and multiplier` for the latter group

acpaquette commented on isis/tests/FunctionalTestsChan1m32isis.cpp:
```diff
@@ -0,0 +1,558 @@
+  Pvl *isisLabel = cube.label();
+
+  // Dimensions Group
+  PvlGroup &dimensions = isisLabel->findGroup("Dimensions", Pvl::Traverse);
+  ASSERT_EQ(int(dimensions["Samples"]), 304);
+  ASSERT_EQ(int(dimensions["Lines"]), 5);
+  ASSERT_EQ(int(dimensions["Bands"]), 3);
+
+  // Pixels Group
+  PvlGroup &pixels = isisLabel->findGroup("Pixels", Pvl::Traverse);
+  ASSERT_EQ(pixels["Type"][0].toStdString(), "Real");
+  ASSERT_EQ(pixels["ByteOrder"][0].toStdString(), "Lsb");
+  ASSERT_EQ(double(pixels["Base"]), 0.0);
+  ASSERT_EQ(double(pixels["Multiplier"]), 1.0);
+
```
For subsequent tests outside of the "default" it maybe unnecessary to check these groups as it's often the same for images from the same instrument.

acpaquette commented on isis/tests/FunctionalTestsChan1m32isis.cpp:
```diff
@@ -0,0 +1,558 @@
+  PvlGroup &kernel = isisLabel->findGroup("Kernels", Pvl::Traverse);
+  ASSERT_EQ(int(kernel["NaifFrameCode"]), -86520);
+
+  Histogram *hist = cube.histogram();
```
Convert these to a smart pointer `std::unique_ptr<Histogram> hist (cube->histogram());` and the deletes can be removed.

acpaquette commented on isis/tests/FunctionalTestsChan1m32isis.cpp:
```diff
@@ -0,0 +1,558 @@
+  ASSERT_EQ(double(pixels["Multiplier"]), 1.0);
+
+  // Instrument Group
+  PvlGroup &inst = isisLabel->findGroup("Instrument", Pvl::Traverse);
+  ASSERT_EQ(inst["SpacecraftName"][0].toStdString(), "CHANDRAYAAN-1");
+  ASSERT_EQ(inst["InstrumentId"][0].toStdString(), "M3" );
+  ASSERT_EQ(inst["TargetName"][0].toStdString(), "MOON" );
+  ASSERT_EQ(inst["SpacecraftClockStartCount"][0].toStdString(), "2/1531046.542" );
+  ASSERT_EQ(inst["SpacecraftClockStopCount"][0].toStdString(), "2/1531047.050" );
+  ASSERT_DOUBLE_EQ(double(inst["LineExposureDuration"]), 101.76);
+  ASSERT_EQ(inst["StartTime"][0].toStdString(), "2008-11-29T17:14:29.729807" );
+  ASSERT_EQ(inst["StopTime"][0].toStdString(), "2008-11-29T17:14:30.238607" );
+  ASSERT_EQ(inst["SpacecraftYawDirection"][0].toStdString(), "FORWARD" );
+  ASSERT_EQ(inst["OrbitLimbDirection"][0].toStdString(), "DESCENDING" );
```
Similarly outside of the default I don't know if it's worth checking every keyword here. This test really only seems to care that `SpacecraftYawDirection` is FORWARD and `OrbitLimbDirection` is DESCENDING. For tests that are explicitly looking to test those features I don't think it's super important to test all keywords in the instrument group.

acpaquette commented on isis/tests/FunctionalTestsChan1m32isis.cpp:
```diff
@@ -0,0 +1,558 @@
+
+  // BandBin Group
+  // Check size, first, 2 middle, and last values? Enough?
+  PvlGroup &bandbin = isisLabel->findGroup("BandBin", Pvl::Traverse);
+  ASSERT_EQ(bandbin["Center"].size(), 85);
+  ASSERT_EQ(bandbin["FilterNumber"].size(), 85);
+  ASSERT_EQ(bandbin["OriginalBand"].size(), 85);
+
+  ASSERT_DOUBLE_EQ(bandbin["Center"][0].toDouble(), 460.990);
+  ASSERT_DOUBLE_EQ(bandbin["Center"][21].toDouble(), 1009.95);
+  ASSERT_DOUBLE_EQ(bandbin["Center"][42].toDouble(), 1429.15);
+  ASSERT_DOUBLE_EQ(bandbin["Center"][84].toDouble(), 2976.20);
+
+  ASSERT_DOUBLE_EQ(bandbin["FilterNumber"][0].toDouble(), 5);
+  ASSERT_DOUBLE_EQ(bandbin["FilterNumber"][21].toDouble(), 57);
+  ASSERT_DOUBLE_EQ(bandbin["FilterNumber"][42].toDouble(), 99);
+  ASSERT_DOUBLE_EQ(bandbin["FilterNumber"][84].toDouble(), 253);
+
+  ASSERT_DOUBLE_EQ(bandbin["OriginalBand"][0].toDouble(), 1);
+  ASSERT_DOUBLE_EQ(bandbin["OriginalBand"][21].toDouble(), 22);
+  ASSERT_DOUBLE_EQ(bandbin["OriginalBand"][42].toDouble(), 43);
+  ASSERT_DOUBLE_EQ(bandbin["OriginalBand"][84].toDouble(), 85);
```
Similar thought with the bandbins group. I don't think it's super important to test outside of the default test but I would like to hear other opinions/points regarding this.

acpaquette reviewed (commented):
A few things to address regarding each test written for this instrument

AustinSanders committed:
Addresses PR feedback

Copies loc/obs files to temp directory
Converts hist pointer to smart pointer
Removes some unnecessary checks for keys in each test
Uses Cube functions instead of label for samples, lines, bands, etc.

acpaquette reviewed (dismissed):
Looks good!

AustinSanders committed:
Replaced badfile apollo with kaguya
